### PR TITLE
Code review

### PR DESCRIPTION
**Pull request by Yash-bhagwat**
This code-review branch contains an extra main.rs whose contents can be reviewed using this pull-request. Thanks!

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    concept::{Attribute, Concept, Value},
+    Connection, DatabaseManager, Session,
+    SessionType::{Data, Schema},
+    TransactionType::{Read, Write}
+};
+use futures::{StreamExt};
+use  std::fs;
+use text_io::read;
+
+const TEST_DATABASE: &str = "codeforces-data-model";
```
Comment (nikolaimerritt) [inline src/main.rs:18]: It's great that you've made this a constant -- it makes the code easier to read.

File: src/main.rs [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+ 
+fn new_core_connection() -> typedb_client::Result<Connection> {
+    Connection::new_plaintext("localhost:1729")
+}
+
+async fn load_data(connection: Connection)->std::io::Result<()>{
```
Comment (nikolaimerritt) [inline src/main.rs]: In Rust, it's best if you try to use `.unwrap()` as little as possible. Instead, try to return errors (`Err(...)`). That way your code will crash less. And you can decide what to do if your functions returns an error. 

Your code encounters two types of errors: 
- `std::io::Error` and 
- `typedb_client::error::Error`. 

You can define an enum that wraps both of these errors. When you encounter an error, wrap the error in your enum. An example is here: https://doc.rust-lang.org/rust-by-example/error/multiple_error_types/wrap_error.html

File: src/main.rs [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    let transaction = session.transaction(Write).await.unwrap();
+    let _ = transaction.query().insert(data.as_str());
+    transaction.commit().await.unwrap();
+    println!("\nData Loaded Successfully\n");
+    Ok(())
+}
+
+async fn load_schema(connection: Connection)->std::io::Result<()>{
+    let schema = fs::read_to_string("./src/schema.tql")?;
+    // query_options(con.clone()).await.unwrap();
```
Comment (nikolaimerritt) [inline src/main.rs]: Can you clean up this comment?

File: src/main.rs [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+    let con=new_core_connection().expect("Line: 74");
+    load_schema(con.clone()).await?;
+    run_io(con.clone()).await;
+
+    Ok(())
+}
+
+#[tokio::main]
+async fn main(){
```
Comment (nikolaimerritt) [inline src/main.rs]: `main` can return a `Result`. So you don't need to create a separate `mymain` function.

File: src/main.rs [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    // query_options(con.clone()).await.unwrap();
+    let databases = DatabaseManager::new(connection.clone());
+    if databases.contains(TEST_DATABASE).await.unwrap()==false {
+        let _ = databases.create(TEST_DATABASE).await;
+        // define schema
+        let session = Session::new(databases.get(TEST_DATABASE).await.unwrap(), Schema).await.unwrap();
+        let transaction = session.transaction(Write).await.unwrap();
+        transaction.query().define(schema.as_str()).await.unwrap();
+        transaction.commit().await.unwrap();
+        session.force_close().unwrap();
```
Comment (nikolaimerritt) [inline src/main.rs]: You shouldn't need to call `.force_close()` here. Sessions should close themselves once dropped.

File: src/main.rs [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+    full_query = full_query + query_part_1 + x.as_str() + query_part_2;
+
+    println!("{}", full_query);
+
+    let mut answer_stream = transaction.query().match_(&full_query.as_str()).unwrap();
+    while let Some(result) = answer_stream.next().await{
+        match result {
+            Ok(concept_map) => {
+                for (_, concept) in concept_map {
+                    if let Concept::Attribute(Attribute { value: Value::String(value), .. }) = concept {
```
Comment (nikolaimerritt) [inline src/main.rs]: Very nice use of parameter destructuring!

File: src/main.rs [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+async fn query2(connection: Connection)->std::io::Result<()>{    
+    let databases = DatabaseManager::new(connection.clone());
+    let session = Session::new(databases.get(TEST_DATABASE).await.unwrap(), Data).await.unwrap();
+    let transaction = session.transaction(Read).await.unwrap();
+    println!("Query chosen : Get IDs of problems with a particular tag");
+    println!("Choose tag");
+    let x: String=read!("{}\n");
+    let query_part_1 ="match $tt ($x, $y) isa possesses-tag; $x isa problem, has problem-number $a; $y isa topic, has topic-name \"";
```
Comment (nikolaimerritt) [inline src/main.rs]: You can use `format!` to insert the values into strings. An example is here: https://doc.rust-lang.org/std/macro.format.html

**Review by nikolaimerritt — CHANGES_REQUESTED** [review]

File: src/main.rs (hunk 2 above)
Comment (Yash-bhagwat) [reply src/main.rs]: Okay sure, thanks for the suggestion!

File: src/main.rs (hunk 3 above)
Comment (Yash-bhagwat) [reply src/main.rs]: Fixed! Unfortunately, missed my eye!

File: src/main.rs (hunk 4 above)
Comment (Yash-bhagwat) [reply src/main.rs]: Fixed!

File: src/main.rs [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+fn new_core_connection() -> typedb_client::Result<Connection> {
+    Connection::new_plaintext("localhost:1729")
+}
+
+async fn load_data(connection: Connection)->Result<(), HandleError>{
+    let data=fs::read_to_string("./src/data.tql").map_err(HandleError::Io)?;    
+    let databases = DatabaseManager::new(connection.clone());
+    let session = Session::new(databases.get(TEST_DATABASE).await.map_err(HandleError::TypeDB)?, Data).await.map_err(HandleError::TypeDB)?;
+    let transaction = session.transaction(Write).await.map_err(HandleError::TypeDB)?;
+    let _ = transaction.query().insert(data.as_str());
```
Comment (nikolaimerritt) [inline src/main.rs]: Why do we assign this value to an unused variable?

File: src/main.rs [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    transaction.commit().await.map_err(HandleError::TypeDB)?;
+    println!("\nData Loaded Successfully\n");
+    Ok(())
+}
+
+async fn load_schema(connection: Connection)->Result<(), HandleError>{
+    let schema = fs::read_to_string("./src/schema.tql").map_err(HandleError::Io)?;
+    let databases = DatabaseManager::new(connection.clone());
+    if databases.contains(TEST_DATABASE).await.map_err(HandleError::TypeDB)?==false {
```
Comment (nikolaimerritt) [inline src/main.rs]: Can we make this `if`/`else` check more consise, rather than checking for `== false` ?

File: src/main.rs [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+    let databases = DatabaseManager::new(connection.clone());
+    let session = Session::new(databases.get(TEST_DATABASE).await.map_err(HandleError::TypeDB)?, Data).await.map_err(HandleError::TypeDB)?;
+    let transaction = session.transaction(Write).await.map_err(HandleError::TypeDB)?;
+    let _ = transaction.query().insert(data.as_str());
+    transaction.commit().await.map_err(HandleError::TypeDB)?;
+    println!("\nData Loaded Successfully\n");
+    Ok(())
+}
+
+async fn load_schema(connection: Connection)->Result<(), HandleError>{
```
Comment (nikolaimerritt) [inline src/main.rs]: The name `load_schema` is a bit misleading, because this function also loads the data into the database.

File: src/main.rs [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+        match result {
+            Ok(concept_map) => {
+                for (_, concept) in concept_map {
+                    if let Concept::Attribute(Attribute { value: Value::String(value), .. }) = concept {
+                        println!("{}",value);
+                    }
+                }
+            }
+            Err(err) => {
+                panic!("An error occurred fetching answers of a Match query: {err}")
```
Comment (nikolaimerritt) [inline src/main.rs]: Can we return an error here, rather than panic?

**Review by nikolaimerritt — CHANGES_REQUESTED** [review]

File: src/main.rs (hunk 9 above)
Comment (Yash-bhagwat) [reply src/main.rs]: Switching the if else condition so as to avoid this ugly " == false " condition

File: src/main.rs (hunk 10 above)
Comment (Yash-bhagwat) [reply src/main.rs]: Changed it to load schema and data

File: src/main.rs [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+    Ok(input_string)
+}
+
+async fn load_data(connection: Connection)->Result<(), HandleError>{
+    let data=fs::read_to_string("./src/data.tql").map_err(HandleError::Io)?;    
+    let databases = DatabaseManager::new(connection.clone());
+    let session = Session::new(databases.get(TEST_DATABASE).await.map_err(HandleError::TypeDB)?, Data).await.map_err(HandleError::TypeDB)?;
+    let transaction = session.transaction(Write).await.map_err(HandleError::TypeDB)?;
+    let inserted_query = transaction.query().insert(data.as_str()).map_err(HandleError::TypeDB)?;
+    let _ = inserted_query.try_collect::<Vec<_>>().await;
```
Comment (nikolaimerritt) [inline src/main.rs]: Can we handle the error returned by `.try_collect`? If we do so, we shouldn't need to assign the return value to a variable.

**Review by nikolaimerritt — CHANGES_REQUESTED** [review]

File: src/main.rs [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+fn read_input() -> Result<String, HandleError>{
+    let mut input_string = String::new();
+    io::stdin().read_line(&mut input_string).map_err(HandleError::Io)?;
+    input_string.pop();
+    Ok(input_string)
+}
+
+async fn load_data(connection: Connection)->Result<(), HandleError>{
```
Comment (nikolaimerritt) [inline src/main.rs]: Can we use a more consistent code formatting? For instance, the arrows between the function declaration and the return type have inconsistent spacing. We'd recommend you to use the [`rustfmt`](https://github.com/rust-lang/rustfmt) tool to format your code in a consistent way.

**Review by nikolaimerritt — CHANGES_REQUESTED** [review]

File: src/main.rs [hunk 14]
```diff
@@ -16,41 +16,82 @@ enum HandleError {
...
+    io::stdin()
+        .read_line(&mut input_string)
+        .map_err(HandleError::Io)?;
     input_string.pop();
     Ok(input_string)
 }
 
-async fn load_data(connection: Connection)->Result<(), HandleError>{
-    let data=fs::read_to_string("./src/data.tql").map_err(HandleError::Io)?;    
+async fn load_data(connection: Connection) -> Result<(), HandleError> {
```
Comment (nikolaimerritt) [inline src/main.rs:33]: This code format looks great!

File: src/main.rs [hunk 15]
```diff
@@ -60,22 +101,40 @@ async fn load_schema_and_data(connection: Connection)->Result<(), HandleError>{
     Ok(())
 }
 
-async fn query1(connection: Connection)->Result<(), HandleError>{
+async fn query1(connection: Connection) -> Result<(), HandleError> {
```
Comment (nikolaimerritt) [inline src/main.rs]: Can we give the `query1`, `query2` and `query3` functions more descriptive names? Ideally reading the function name would tell us what kind of query these functions execute.

**Review by nikolaimerritt — CHANGES_REQUESTED** [review]

File: src/main.rs [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+        .await
+        .map_err(HandleError::TypeDB)?
+    {
+        println!("\nSchema Already Defined\n");
+    } else {
+        databases
+            .create(TEST_DATABASE)
+            .await
+            .map_err(HandleError::TypeDB)?;
+        // define schema
```
Comment (nikolaimerritt) [inline src/main.rs]: Could we please remove this comment? It is clear from the code that the schema is loaded here.

**Review by nikolaimerritt — APPROVED** [review]

File: src/main.rs (hunk 16 above)
Comment (Yash-bhagwat) [reply src/main.rs]: Sure! Done so!

**Pull request merged**